llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp

```diff
@@ -626,6 +626,127 @@ Instruction *InstCombinerImpl::foldPowiReassoc(BinaryOperator &I) {
   return nullptr;
 }
 
+bool isFSqrtDivToFMulLegal(Instruction *X, SmallSetVector<Instruction *, 2> &R1,
+                           SmallSetVector<Instruction *, 2> &R2) {
+
+  BasicBlock *BBx = X->getParent();
+  BasicBlock *BBr1 = R1[0]->getParent();
+  BasicBlock *BBr2 = R2[0]->getParent();
+
+  auto IsStrictFP = [](Instruction *I) {
+    IntrinsicInst *II = dyn_cast<IntrinsicInst>(I);
+    return II && II->isStrictFP();
+  };
+
+  // Check the constaints on instruction X.
+  auto XConstraintsSatisfied = [X, &IsStrictFP]() {
+    if (IsStrictFP(X))
+      return false;
+    // X must atleast have 4 uses.
+    // 3 uses as part of
+    //    r1 = x * x
+    //    r2 = a * x
+    // Now, post-transform, r1/r2 will no longer have usage of 'x' and if the
+    // changes to 'x' need to persist, we must have one more usage of 'x'
+    if (!X->hasNUsesOrMore(4))
+      return false;
+    // Check if reciprocalFP is enabled.
+    bool RecipFPMath = dyn_cast<FPMathOperator>(X)->hasAllowReciprocal();
```

Comment:
`arcp` is not the right fast-math flag here. You probably want `reassoc`.

Reply:
I disagree here. What  are we re-associating here? Nothing

On the other hand, we are transforming 
 x = 1/sqrt(a)                   ( i.e. = sqrt(a)/a  )
TO
x = sqrt(a)  * 1/a
This is clearly reciprocal math optimization as per the clang docs.

Are you hinting at reassoc flag somewhere else?

Reply:
The definition of most of our fast-math flags are about as clear as mud. However, the `arcp` flag is generally treated as allowing the transformation `x / y` => `x * (1.0 / y)`, and only that transformation.

`1.0 / sqrt(a)` is *not* the same as `sqrt(a) / a`--it is for real numbers, but FP numbers are not real numbers--so converting `1.0 / sqrt(a)` to `sqrt(a) / a` requires a fast-math flag of some kind, and the one that we've been abusing here is the `reassoc` flag, which has kind of devolved in practice into "permit algebraic rewrites".

Reply:
> `1.0 / sqrt(a)` is _not_ the same as `sqrt(a) / a`--it is for real numbers, but FP numbers are not real numbers--so converting `1.0 / sqrt(a)` to `sqrt(a) / a` requires a fast-math flag of some kind, and the one that we've been abusing here is the `reassoc` flag, which has kind of devolved in practice into "permit algebraic rewrites".

So, do you suggest that  `1.0 / sqrt(a)` to `sqrt(a) / a` requires reassoc flag and then convert `sqrt(a) / a` to `sqrt(a) * 1/a`   under arcp flag?
If my understanding is correct, this would require combination of `reassoc arcp`, right?

Reply:
ping